### ISSUE TITLE
LINQ implementation chokes on interface member but not on same member in implementation

### ISSUE THREAD

**Issue by oliverw** [Azure/azure-mobile-apps#627]
## Describe the bug

Interface

```c#
public interface IDatasyncEntity
{
    string Id { get; set; }
    bool Deleted { get; set; }
    DateTimeOffset UpdatedAt { get; set; }
    string? Version { get; set; }
}
```

Implementation

```c#
public class ReactiveDatasyncEntity : ReactiveObject,
    IDatasyncEntity
{
    [JsonProperty("id")]
    public string Id { get; set; } = null!;

    [JsonProperty("deleted")]
    public bool Deleted { get; set; }
        
    [JsonProperty("updatedAt")]
    public DateTimeOffset UpdatedAt { get; set; }
        
    [JsonProperty("version")]
    public string? Version { get; set; }
}
```

Usage 

```c#
private async Task SyncTableAsync<T>(IOfflineTable<T> table, CancellationToken ct)
    where T: IDatasyncEntity
{
    var deletedItemsQuery = table.CreateQuery().Where(x => x.Deleted);

    await table.PurgeItemsAsync(deletedItemsQuery, new PurgeOptions
    {
        DiscardPendingOperations = true
    }, ct);
}
```

This crashes with: _The member 'Deleted is not supported in the 'Where' clause_

>    at Microsoft.Datasync.Client.Query.Linq.FilterBuildingExpressionVisitor.VisitMemberExpression(MemberExpression node)
>    at Microsoft.Datasync.Client.Query.Linq.FilterBuildingExpressionVisitor.Visit(Expression node)
>    at Microsoft.Datasync.Client.Query.Linq.FilterBuildingExpressionVisitor.Compile(Expression expression, DatasyncContractResolver contractResolver)
>    at Microsoft.Datasync.Client.Query.Linq.QueryTranslator`1.AddFilter(MethodCallExpression expression)
>    at Microsoft.Datasync.Client.Query.Linq.QueryTranslator`1.VisitMethodCall(MethodCallExpression expression)
>    at Microsoft.Datasync.Client.Query.Linq.QueryTranslator`1.Translate()
>    at Microsoft.Datasync.Client.Query.TableQuery`1.ToODataString(Boolean includeParameters)
>    at Microsoft.Datasync.Client.Table.OfflineTable`1.PurgeItemsAsync[U](ITableQuery`1 query, PurgeOptions options, CancellationToken cancellationToken)
>    at Done.Core.Services.DataSyncService.SyncTableAsync[T](IOfflineTable`1 table, CancellationToken ct)
>    at Done.Core.Services.DataSyncService.SyncTablesAsync(SyncRequest request, CancellationToken ct)
> 

This works (diffent generic constraint).

```c#
private async Task SyncTableAsync<T>(IOfflineTable<T> table, CancellationToken ct)
    where T: Model.ReactiveDatasyncEntity
{
    var deletedItemsQuery = table.CreateQuery().Where(x => x.Deleted);

    await table.PurgeItemsAsync(deletedItemsQuery, new PurgeOptions
    {
        DiscardPendingOperations = true
    }, ct);
}
```

## What platforms?

* Client (.NET, JavaScript, iOS, Android):
  * .NET 6
  * 5.1.1
  * Windows 11

**Comment by adrianhall** [comment]
Interesting - will set up a test case for this (as it's non-obvious as to why this is happening.  I take it that the PurgeItemsAsync() is a misnomer - i.e. you can replicate with a CreateQuery.Where(...) where T is an interface, without calling anything?

**Comment by oliverw** [comment]
```c#
  private void FooAsync<T>(IOfflineTable<T> table, CancellationToken ct)
      where T : Model.IDatasyncEntity
  {
      var deletedItemsQuery = table.CreateQuery().Where(x => x.Deleted);
  }

FooAsync(Tasks, ct);
```

Yep, this does not crash.

**Comment by adrianhall** [comment]
Yeah - LINQ is lazy evaluated.  When we do a PurgeItemsAsync(), we do a query.ToODataString() and then convert back to the internal representation (highly inefficient for what it does, and something we can fix in 6.0.0).  I suspect the ToODataString() which effects the JIT evaluation is not working properly.  I'll have to write some tests for it first.

**Comment by adrianhall** [comment]
The following test fails:

```csharp
        [Fact]
        public void Linq_DerivedTest()
        {
            var client = GetMockClient();
            var table = client.GetRemoteTable<DerivedTestDatasyncEntity>("derived");
            var queryString = GetDerivedQuery(table);
            Assert.Equal("deleted eq true", queryString);
        }

        public static string GetDerivedQuery<T>(IRemoteTable<T> table) where T : IDatasyncEntity
        {
            return table.Where(x => x.Deleted).ToODataString();
        }
```

The root cause of this was in the contract resolver - the expression is a MemberExpression, but derived type get called with NodeType == Convert instead of NodeType == Parameter.  As a result, I added a second test for NodeType == Convert && member.MemberType == Property.  This seems to solve the issue.
